galsim/fitswcs.py

```diff
@@ -1761,3 +1763,92 @@ def FitsWCS(file_name=None, dir=None, hdu=None, header=None, compression='auto',
 FitsWCS._takes_rng = False
 
 
+def FittedSIPWCS(x, y, ra, dec, order=3, center=None):
+    """A WCS constructed from a list of reference celestial and image
+    coordinates.
+
+    Parameters:
+        x:      Image x-coordinates of reference stars in pixels
+        y:      Image y-coordinates of reference stars in pixels
+        ra:     Right ascension of reference stars in radians
+        dec:    Declination of reference stars in radians
+        order:  The order of the Simple Imaging Polynomial (SIP) used to
+                describe the WCS distortion.  Should be in the range [1, 9]
+                [default: 3]
+        center: A `CelestialCoord` defining the location on the sphere where the
+                tangent plane is centered.  [default: None, which means use the
+                average position of the list of reference stars]
+    """
+    from scipy.optimize import least_squares
+    from .celestial import CelestialCoord
+    if center is None:
+        # Use deprojected 3D mean of ra/dec unit sphere points as center
+        wx = np.mean(np.cos(dec)*np.cos(ra))
+        wy = np.mean(np.cos(dec)*np.sin(ra))
+        wz = np.mean(np.sin(dec))
+        center = CelestialCoord.from_xyz(wx, wy, wz)
+
+    if order < 1 or order > 9:
+        raise GalSimValueError("Illegal SIP order", order)
```

Comment:
Yeah, on page 2 [here](https://fits.gsfc.nasa.gov/registry/sip/SIP_distortion_v1_0.pdf) there's a statement: "A_ORDER and B_ORDER can take on integer values ranging from 2 to 9".  But now that you mention it, I can't see any technical reason these couldn't be larger.  I guess maybe we should allow order > 9, but issue a warning if writing these SIP coefficients to a header since other software may not be able to read them?

And `order=1` just means don't do the SIP stuff.  I should double check that the header writer isn't inserting 'TAN_SIP' or 'A_ORDER', 'B_ORDER' in this case though.

Reply:
I don't see a good reason for a limit < 100.  At 100, the FITS header keys would start to be more than 8 characters, which means they will require HEIRARCH cards, so bleh.  But <= 9 seems a bit arbitrary and unnecessary.

Reply:
I went ahead and removed this restriction.  To be fair though, I don't think I'd recommend setting `order` this high.  I believe every fit I looked at actually landed on a better solution when order=7 than when order=9.  So either there are numerical difficulties with high-order non-centered, non-Chebyshev/Legendre polynomials (definitely the case in the inverse case, where it's possible to end up with situations like dx~1000 pixels raised to the 9th power say) or maybe there are just more local minima to get trapped in.  I guess we could gradually increase the order when fitting to at least obtain monotonically improving results, but I haven't bothered here since the smallish field-of-view examples all do well with order=3, including LSST sims.

Reply:
Sounds good.  9 in the SIP spec might just have been this kind of practical limitation rather than something fundamental.  As you say, it's probably a bad idea to go that high anyway.